File: game/bird.py
import pygame
import numpy as np

class Bird:
    def __init__(self, x, y, index):
        self.x = x
        self.y = y
        self.index = index
        
        # Physics - adjusted for better control
        self.velocity = 0
        self.gravity = 0.4  # Reduced further
        self.flap_strength = -7  # Reduced further
        self.max_velocity = 8  # Add velocity cap
        
        # Dimensions
        self.width = 34
        self.height = 24
        
        # Game stats
        self.alive = True
        self.fitness = 0
        self.score = 0
        self.distance = 0
        self.last_pipe_x = float('inf')
        
        # Previous generation stats
        self.fitness_prev = 0
        self.score_prev = 0
        
    def flap(self):
        if self.alive:
            self.velocity = self.flap_strength
    
    def update(self):
        if not self.alive:
            return
            
        # Apply gravity
        self.velocity += self.gravity
        # Cap velocity
        self.velocity = min(max(self.velocity, -self.max_velocity), self.max_velocity)
        self.y += self.velocity
        
        # Update distance
        self.distance += 1
        
    def draw(self, screen):
        if not self.alive:
            return
            
        color = (255, 255, 0) if self.alive else (150, 150, 150)
        pygame.draw.rect(screen, color, (self.x, self.y, self.width, self.height))
    
    def die(self):
        self.alive = False
        
    def get_rect(self):
        return pygame.Rect(self.x, self.y, self.width, self.height) 

File: static/js/bird.js
class Bird {
    constructor(x, y, index) {
        this.x = x;
        this.y = y;
        this.index = index;
        
        this.gravity = 0.4;         // Increased from 0.35
        this.flapStrength = -6.5;   // Adjusted from -5.5
        this.maxVelocity = 10;      // Increased from 8
        this.terminalVelocity = 15; // Increased from 12
        this.smoothing = 0.8;       // Reduced from 0.9 for more responsive movement
        
        // Dimensions
        this.width = 34;
        this.height = 24;
        
        // Game stats
        this.alive = true;
        this.fitness = 0;
        this.score = 0;
        this.distance = 0;
        this.lastPipeX = 0;
        
        // Previous gen stats
        this.fitnessPrev = 0;
        this.scorePrev = 0;
        
        // Neural network related properties
        this.currentInputs = null;
        this.currentOutput = null;
        
        // Initialize velocity
        this.velocity = 0;
        
        // Load image with proper error handling
        this.image = new Image();
        this.image.crossOrigin = "anonymous";
        this.imageLoaded = false;
        
        // Set onload before src
        this.image.onload = () => {
            this.imageLoaded = true;
            console.log('Bird image loaded successfully');
        };
        
        this.image.onerror = (error) => {
            console.error('Failed to load bird image:', error);
            this.imageLoaded = false;
        };
        
        // Set src last
        this.image.src = '/static/assets/flappy-bird.png';
    }
    
    flap() {
        if (this.alive) {
            this.velocity = this.flapStrength;
        }
    }
    
    update() {
        if (!this.alive) return;
        
        // Apply gravity
        this.velocity += this.gravity;
        
        // Clamp velocity
        this.velocity = Math.max(
            Math.min(this.velocity, this.terminalVelocity),
            -this.maxVelocity
        );
        
        // Update position
        this.y += this.velocity;
        
        // Update distance
        this.distance += 0.5;
        
        // Keep bird in bounds
        if (this.y < 0) {
            this.y = 0;
            this.velocity = 0;
        }
    }
    
    draw(ctx) {
        if (!this.alive || !this.imageLoaded) return;
        
        // Draw only the bird image, remove debug box
        ctx.drawImage(this.image, this.x, this.y, this.width, this.height);
        
        // Remove any debug visualization code
    }
    
    die() {
        this.alive = false;
    }
    
    getRect() {
        return {
            x: this.x,
            y: this.y,
            width: this.width,
            height: this.height
        };
    }
} 